pkg/api/validation/validation.go

```diff
@@ -1278,6 +1278,37 @@ func validatePullPolicy(policy api.PullPolicy, fldPath *field.Path) field.ErrorL
 	return allErrors
 }
 
+func validateInitContainers(containers, otherContainers []api.Container, volumes sets.String, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(containers) > 0 {
+		allErrs = append(allErrs, validateContainers(containers, volumes, fldPath)...)
+	}
+
+	allNames := sets.String{}
+	for _, ctr := range otherContainers {
+		allNames.Insert(ctr.Name)
+	}
+	for i, ctr := range containers {
+		idxPath := fldPath.Index(i)
+		if allNames.Has(ctr.Name) {
+			allErrs = append(allErrs, field.Duplicate(idxPath.Child("name"), ctr.Name))
+		}
+		if len(ctr.Name) > 0 {
+			allNames.Insert(ctr.Name)
+		}
+		if ctr.Lifecycle != nil {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("lifecycle"), ctr.Lifecycle, "must not be set for init containers"))
+		}
+		if ctr.LivenessProbe != nil {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("livenessProbe"), ctr.LivenessProbe, "must not be set for init containers"))
+		}
+		if ctr.ReadinessProbe != nil {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("readinessProbe"), ctr.ReadinessProbe, "must not be set for init containers"))
```

Comment:
Originally was going to remove, decided that we could tolerate them being
attachable.

On Fri, Apr 8, 2016 at 4:41 PM, Prashanth B <notifications@github.com>
wrote:

> In pkg/api/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/23567#discussion_r59086228>
> :
>
> > +	for i, ctr := range containers {
> > +		idxPath := fldPath.Index(i)
> > +		if allNames.Has(ctr.Name) {
> > +			allErrs = append(allErrs, field.Duplicate(idxPath.Child("name"), ctr.Name))
> > +		}
> > +		if len(ctr.Name) > 0 {
> > +			allNames.Insert(ctr.Name)
> > +		}
> > +		if ctr.Lifecycle != nil {
> > +			allErrs = append(allErrs, field.Invalid(idxPath.Child("lifecycle"), ctr.Lifecycle, "must not be set for init containers"))
> > +		}
> > +		if ctr.LivenessProbe != nil {
> > +			allErrs = append(allErrs, field.Invalid(idxPath.Child("livenessProbe"), ctr.LivenessProbe, "must not be set for init containers"))
> > +		}
> > +		if ctr.ReadinessProbe != nil {
> > +			allErrs = append(allErrs, field.Invalid(idxPath.Child("readinessProbe"), ctr.ReadinessProbe, "must not be set for init containers"))
>
> also stdin, tty, stdinonce? though I suppose it would be nice to have a
> -it that dropped you into an init container so you can eg: write out
> nginx.conf before nginx starts
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23567/files/53b5dd4d022c2b97f02e40d07c81600fc249b5ae..c75dc6a7e701f3e4484e6545658cf85711a2d6d7#r59086228>
>